src/gc/objecthandle.cpp

```diff
@@ -285,7 +285,7 @@ void CALLBACK AsyncPinObject(_UNCHECKED_OBJECTREF *pObjRef, uintptr_t *pExtraInf
     Object **pRef = (Object **)pObjRef;
     _ASSERTE(lp2);
     promote_func* callback = (promote_func*)lp2;
-    callback(pRef, (ScanContext *)lp1, GC_CALL_PINNED);
```

Comment:
@Maoni0 @davmason This is changing contract of the local GC. Do I need to worry about it?

Reply:
I don't see how this affects LocalGC... it's just making a callback that's implemented internally by the GC. 

sorry I haven't had time to look at the rest of the changes - do you mind explaining why you are not passing the GC_CALL_PINNED flag anymore? if the object is now a native obj, it wouldn't need to call the callback at all - GC is not going to do anything with the object. do we still have Overlapped managed objects at all anymore (or did you split up the native part so the managed part doesn't need to be pinned)? if not then do we still need this async pinned handle type?

Reply:
We still have Overlapped managed objects with this change, but the ASYNC handles do not pin them anymore - only the data arrays that they are pointing to are pinned by the ASYNC handles with this change.

It is the change in the GC contract: the ASYNC handles were pinning the overlapped objects before this change and they are not pinning the overlapped objects after this change.

Reply:
ok then if you run with this version of clrgc.dll with an older version of clr.dll it wouldn't work. but this doesn't have to be a breaking change - we could communicate the version of the clr.dll to clrgc.dll and have async pinned handles do different things based on the minor version. but of course this requires some changes in LocalGC. what do you think, @davmason? otherwise you'd need to change the major version (GC_INTERFACE_MAJOR_VERSION) which is probably ok at this point if we think not many people will be using the 2.1 LocalGC anyway.

I know you see a big improvement in a microbenchmark - in real world scenarios the perf problem that these overlapped structures introduce is always about fragmentation and a lot of the fragmentation comes from the data that gets pinned so I am not sure how much improvement you'd see there.

Reply:
We probably need to rev the major version for 2.2 anyways, so I think we can just accept that and make the 2.2 local gc incompatible with 2.1. There are some other changes too that likely need an api change for 2.2 (like the debugger notifications in #17769). If this were the only issue it might make sense to talk about ways to avoid it, but it seems inevitable at this point.

Reply:
My primary motivation for this change has been code simplification. I have done the performance measurements just to ensure I am not regressing performance. The improvement was a nice surprise.

I have reved the GC interface version.